[Language Server] Add "create undefined function" code action support to language server

IMS94 opened the pull request:
## Purpose
Add "create undefined function" code action support for basic scenarios.

Fixes #27424

## Approach
Works for 
- Simple function calls:
```
   addTwoIntegers(a, c, true);
```
- Function calls in var declarations:
```
   int d = addTwoIntegers2(a, b);
```
- Function calls with parameters of different modules:
```
    module1:Student student = createStudent(name, age);
    addStudent(student);
    module1:Student converted = convertStudent(student);
```
Doesn't work for `assignment statements`, `let expressions`, assignments within record types, etc yet.

## Samples
- `addTwoIntegers(a, c, true);` will result:
```
function addTwoIntegers(int a, float b, boolean c) {
    
}
```

- `module1:Student student = createStudent(name, age);` will result in a function like:
```
function createStudent(string a, any b) returns module1:Student {
    
    return {};
}
```
- `int d = addTwoIntegers2(a, b);` will result in:
```
function addTwoIntegers2(int a, int b) returns int {
    
    return 0;
}
```
and etc.

## Remarks
> List any other known issues, related PRs, TODO items, or any other notes related to the PR.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

rasika commented on language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/CreateFunctionCodeAction.java:
```diff
@@ -73,8 +76,26 @@ public boolean isEnabled(LanguageServerContext serverContext) {
+            }
+        }
+        
+        if (callExpr != null) {
```
we can remove this redundant check with early exit

nadeeshaan commented on language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/command/executors/CreateFunctionExecutor.java:
```diff
@@ -89,153 +97,99 @@ public Object execute(ExecuteCommandContext context) throws LSCommandExecutorExc
-        if (matchedTypeSymbol == null) {
+
+        FunctionCallExpressionTypeFinder typeFinder = new FunctionCallExpressionTypeFinder(semanticModel);
+        TypeSymbol returnTypeSymbol = typeFinder.typeOf(fnCallExprNode).orElse(null);
```
Let's avoid nulls as much as possible. Shall we use the optional here?

rasika commented on language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/command/visitors/FunctionCallExpressionTypeFinder.java:
```diff
@@ -0,0 +1,120 @@
+ * best to return a valid {@link TypeSymbol}. Else, the returned result will probably be a
+ * {@link TypeDescKind#COMPILATION_ERROR}. Rarely, will return {@code null}.
+ */
+public class FunctionCallExpressionTypeFinder extends NodeVisitor {
```
Let's put `@since 2.0.0` to mark this as introduced in 2.0.0

rasika commented on language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/command/AbstractCommandExecutionTest.java:
```diff
@@ -0,0 +1,133 @@
+/**
+ * Command Execution Test Cases.
+ *
+ * @since 0.982.0
```
Is this `@since` annotated version correct? Let's put `@since 2.0.0` to mark this as introduced in 2.0.0. Anyway it is okay not to have this for test classes since they are not part of the public apis :)

IMS94 replied:
Since the `@since` tag is not present in other tests as well, I have removed it.

IMS94 commented on language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/CreateFunctionCodeAction.java:
```diff
@@ -49,8 +53,7 @@
 
     @Override
     public boolean isEnabled(LanguageServerContext serverContext) {
```
Done!

IMS94 commented on language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/command/executors/CreateFunctionExecutor.java:
```diff
@@ -89,153 +97,99 @@ public Object execute(ExecuteCommandContext context) throws LSCommandExecutorExc
+
+        Document document = context.workspace().document(filePath.get()).orElseThrow();
+        LinePosition linePosition = LinePosition.from(endLine, endCol);
+        Set<String> visibleSymbolNames = semanticModel.visibleSymbols(document, linePosition).stream()
```
Seems like not. Because we get the `ExecuteCommandContext` here. Not `CodeActionContext`.

nadeeshaan replied:
Then, shall we add the API to the `ExecuteCommandContext` as well? WDYT?

IMS94 replied:
The `CodeActionContext` is extended from `DocumentServiceContext` while `ExecuteCommandContext` is extended from `WorkspaceServiceContext` (which isn't related to a specific file). Therefore, if we are to add the mentioned method, it should be like this:
```
    List<Symbol> visibleSymbols(Path filePath, Position position);
```
Will that be fine?

nadeeshaan replied:
Yeah, that would be fine. Id we are gonna invoke the visible symbols, API multiple times within a given operation flow, let's add the API. Otherwise let's use your approach for the moment

IMS94 replied:
Added it to the `WorkspacServiceContext`.

nadeeshaan reviewed (commented):
When we generating the signatures,
1. Do we honour the module aliases
2. Do we honour the pre-declared langlibs?

IMS94 commented:
@nadeeshaan It doesn't honor module aliases as of now. Working on it.

IMS94 commented:
@nadeeshaan I have added support for module aliases. And this works with pre-declared lang libs. I have added tests for both scenarios.

Overall, this implementation works for simple name references, assignment statements and variable declarations. Complex scenarios like let expressions, function calls when creating records, methods inside classes/services are not supported yet.

Furthermore, in the following scenario, the `typeSymbol.signature()` returns `{ string recField; anydata...; }` while the expected value is `{| string recField; anydata...; |}`. Need to address that in a separate issue.
```
record {
      string recField = "";
    } testRec = {};
    record {
      int recField = 1;
    } testRec2 = thisIsANewFunction(testRec);
```

rasika reviewed (approved):
LGTM
